### ISSUE TITLE
TurboQuant decode produces token-0 spam under MTP+PIECEWISE on Qwen3.6-27B

### ISSUE THREAD

**Issue by jasonboukheir** [jasonboukheir/vllm-xpu-nix#84]
Post-#50 (jasonboukheir/vllm#2 / vllm-xpu-nix d83efc4), TurboQuant decode no longer asserts at first decode — but the output it produces is corrupt: the model emits the same low-id token (most often `!`) repeatedly until `max_tokens` is hit. Switching only `kvCacheDtype` to `auto` (bf16, FLASH_ATTN backend), with every other knob held constant, restores coherent output. So the corruption is on the TurboQuant decode path, not in MTP / PIECEWISE / model quantization / cudagraph capture.

## Build

`vllm-0.22.0+unstable.2026.06.01.ge63cd2e-d9993e1d` (jasonboukheir/vllm `main` at `e63cd2ed9`, vllm-xpu-nix `d83efc4`).

## Repro

Brutus chat instance, single curl after the unit reaches `Application startup complete`:

```
curl -s http://127.0.0.1:8000/v1/chat/completions \
  -H content-type:application/json \
  -d '{"model":"qwen3.6-27b",
       "messages":[{"role":"system","content":"You are a helpful assistant."},
                   {"role":"user","content":"Write a haiku about debugging."}],
       "max_tokens":512,"temperature":0}'
```

The only knob swapped between the two runs is `kvCacheDtype`.

### Run A — `kvCacheDtype = "turboquant_k3v4_nc"`, `maxModelLen = 131072`

```json
"reasoning":"!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! ...",
"finish_reason":"length",
"completion_tokens":512
```

(literal `!` for the entire decode horizon)

### Run B — `kvCacheDtype = "auto"`, `maxModelLen = 8192`

```json
"reasoning":"Here's a thinking process:\n\n1.  **Analyze User Request:**\n   - **Topic:** Debugging\n   - **Form:** Haiku (5-7-5 syllable structure)\n   ...\n4.  **Draft - Line 1 (5 syllables):**\n   - Hidden bugs lurk (4) -> Hidden bugs hide (4) -> ... -> *Bugs hide in plain sight* (5) - good\n5.  **Draft - Line 2 (7 syllables):**\n   - Line by line we trace the fault (7) ..."
```

Coherent Qwen3-style chain of thought.

Held constant across A/B: `dtype=bfloat16`, `quantization=inc`, `maxNumSeqs=3`, `gpuMemoryUtilization=0.75`, `speculativeConfig={method:mtp,num_speculative_tokens:2}`, `enforceEager=false`, `enableXpuGraph=true`, `cudagraphMode=PIECEWISE`, `cudagraphCaptureSizes=[3,9]`, `reasoningParser=qwen3`, `enableAutoToolChoice=true`, `toolCallParser=qwen3_xml`, `languageModelOnly=true`.

### Earlier signal that lined up

A short prompt that stops after a handful of tokens *does* produce correct content: `"What is 2+2?"` → `content: "\n\n4"`, finish_reason=stop. So the first decode step or two are right, then the model drifts to the degenerate token. This is consistent with corrupt state accumulating in either the TQ scratch buffers or the TQ KV cache as more decode iterations stack.

## What this isn't

- **Not #50.** EngineCore no longer asserts on first decode. Boot, `/health`, `/v1/models`, and the first decode step all succeed; the issue is the decoded *content*.
- **Not MTP.** Same MTP config produces coherent output under bf16 KV.
- **Not PIECEWISE.** Same `cudagraphMode = "PIECEWISE"` is in both runs.
- **Not INC weight quantization.** Same weights, same `--quantization inc`.

## Hypotheses (turboquant_attn.py, post-fix)

In rough order of suspicion:

1. **Stale-buffer leak across layers / decode steps.** `mid_o_buf` / `lse_buf` are now backed by one global workspace buffer that's reused across all 16 full-attention layers and every decode iteration. The TQ Triton decode kernel almost certainly doesn't zero-initialize inactive kv-split slots in `lse_buf` (every other backend pre-fills with `-inf` at the top of stage 1). If the split count `S` exceeds `ceil(seq_len / kv_block_size)` for the active sequence, the unused splits read leftover values from the previous step — at step N that's step N-1's logits for a different layer / different request. The `!` outputs are what you'd get if `lse_buf` returns stale large values for all splits, collapsing the softmax to a single token.

2. **Reserve shape vs. runtime shape mismatch.** `_reserve_decode_workspace` sizes for `max_batch_tokens = max_num_seqs * (1 + num_speculative_tokens)`. At runtime, decode batches use `B` = active-seqs × verify-shape. For a single-seq request with MTP K=2 that's `B=3`, well under the reserve of 9. `_compute_layout` re-derives offsets per call from the *requested* shape, not the reserved one — so the runtime view starts at offset 0 of `aligned(actual_bytes_for_B=3)`, NOT at the offset reserve baked in. That should be safe (always fits), but worth double-checking that the kernel's strides match what the view reports vs. what it would have under a fresh `torch.empty` of the runtime shape.

3. **Pre-existing TQ decode kernel bug masked by #50.** Before the workspace fix, every first decode died at the assert before any TQ kernel could run, so a quality regression on the decode kernel itself could have been sitting un-exercised. Worth diffing the TQ decode kernel against what older runs (on `Qwen3.6-35B-A3B-GPTQ-Int4`, which #50 noted ran indefinitely) actually executed — if 35B-A3B was using a different code branch (e.g. FusedMoE → different K/V indexing), 27B-dense might be the only path that hits a latent kernel bug.

(1) is the cheapest to verify: add `mid_o_buf.zero_()` / `lse_buf.fill_(float("-inf"))` immediately before the kernel call in `_decode_attention` and re-test the haiku. If output goes coherent, the leak is real and we either pre-zero in the impl or fix the kernel to fully cover all splits.

## Workaround

`services.vllm-xpu.instances.chat.kvCacheDtype = "auto"` — disables TurboQuant attention entirely, falls back to FLASH_ATTN + bf16 KV. Loses the ~32× KV compression so `maxModelLen` drops accordingly (8192 fits cleanly at the current `gpuMemoryUtilization = 0.75`). In place on brutus while this is investigated.

## Why this matters

Qwen3.6-27B + turboquant_k3v4_nc is the whole point of running the int4 weights at 131k context on a 32 GiB B70 — bf16 KV caps usable context at ~8k under the same memory budget. The TQ KV is non-substitutable for that workload (per [feedback_qwen36_is_target_not_substitutable] in our session memory), so this needs a real fix, not just the workaround.
